### PR TITLE
feat: reasoning spec addition

### PR DESCRIPTION
**Pull request by AlemTuzlak**
Implementation for the:
https://docs.ag-ui.com/drafts/reasoning

### CODE REVIEW

File: docs/concepts/events.mdx [hunk 1]
```diff
@@ -553,122 +552,238 @@ implementation across frontends and agents.
...
 
-## Draft Events
+## Reasoning Events
 
-These events are currently in draft status and may change before finalization. They represent proposed extensions to the protocol that are under active development and discussion.
+Reasoning events support LLM reasoning visibility and continuity, enabling
+chain-of-thought reasoning while maintaining privacy. These events allow agents
+to surface reasoning signals (e.g., summaries) and support encrypted reasoning
+items for state carry-over across turns—especially under `store:false` or zero
+data retention policies—without exposing raw chain-of-thought.
```
Comment (tylerslaton) [inline docs/concepts/events.mdx:561]: Could we briefly explain `store:false` or link to something that does?

File: docs/concepts/messages.mdx [hunk 2]
```diff
@@ -107,13 +116,22 @@ interface ToolMessage {
   role: "tool"
   content: string // Result from the tool execution
   toolCallId: string // ID of the tool call this message responds to
+  error?: string // Optional error message if the tool execution failed
+  encryptedValue?: string // Optional encrypted reasoning for state continuity
```
Comment (tylerslaton) [inline docs/concepts/messages.mdx:120]: Would there be a situation where a non `reasoning` role message would contain encrypted content?

File: docs/concepts/messages.mdx [hunk 3]
```diff
@@ -107,13 +116,22 @@ interface ToolMessage {
...
 ```
 
+Key points:
+
+- The `toolCallId` links the result back to the original tool call
+- Use `error` to indicate tool execution failures
+- Use `encryptedValue` to attach encrypted chain-of-thought related to how the
+  agent interpreted or processed the tool result
+
 ### Activity Messages
```
Comment (tylerslaton) [inline docs/concepts/messages.mdx:131]: I could see people confusing `ActivityMessage`s with `ReasoningMessages`. Can we add a [callout](https://www.mintlify.com/docs/components/callouts) that explains the differences and when to use either?

File: docs/concepts/reasoning.mdx [hunk 4]
```diff
@@ -0,0 +1,460 @@
...
+quality. AG-UI's reasoning support addresses three key challenges:
+
+- **Reasoning visibility**: Surface reasoning signals (e.g., summaries) to users
+  without exposing raw chain-of-thought
+- **State continuity**: Maintain reasoning context across turns using encrypted
+  reasoning items, even under `store:false` or zero data retention (ZDR)
+  policies
+- **Privacy compliance**: Support enterprise privacy requirements while
+  preserving reasoning capabilities
+
```
Comment (tylerslaton) [inline docs/concepts/reasoning.mdx:24]: If we do add a [callout](https://www.mintlify.com/docs/components/callouts) above about Reasoning vs Activity, we should do something similar here.

File: docs/concepts/reasoning.mdx [hunk 5]
```diff
@@ -0,0 +1,460 @@
...
+  type: "REASONING_MESSAGE_CONTENT",
+  messageId: "msg-123",
+  delta: "Let me think through this step by step. ",
+}
+
+yield {
+  type: "REASONING_MESSAGE_CONTENT",
+  messageId: "msg-123",
+  delta: "First, I need to analyze the requirements...",
+}
```
Comment (tylerslaton) [inline docs/concepts/reasoning.mdx:177]: Nit pick: I think we should show off an example of a delta here. Since reasoning often streams, that would look something like this.

What do you think?


```suggestion
yield {
  type: "REASONING_MESSAGE_CONTENT",
  messageId: "msg-123",
  delta: "Let me ",
}

yield {
  type: "REASONING_MESSAGE_CONTENT",
  messageId: "msg-123",
  delta: "think through ",
}

yield {
  type: "REASONING_MESSAGE_CONTENT",
  messageId: "msg-123",
  delta: "this step ",
}

yield {
  type: "REASONING_MESSAGE_CONTENT",
  messageId: "msg-123",
  delta: "by step...",
}
```

Comment (tylerslaton) [reply docs/concepts/reasoning.mdx:177]: I'd also be open to showing both to the user. Given some of the ZDR examples, both ways are relevant.

File: docs/sdk/js/core/events.mdx [hunk 6]
```diff
@@ -425,6 +433,162 @@ type CustomEvent = BaseEvent & {
 | `name`   | `string` | Name of the custom event        |
 | `value`  | `any`    | Value associated with the event |
 
+## Reasoning Events
```
Comment (tylerslaton) [inline docs/sdk/js/core/events.mdx:400]: Nit pick: Can we move reasoning above Special events? Since this is a key piece of functionality I think it makes the most sense above the "catch all" events.

File: sdks/python/ag_ui/core/events.py [hunk 7]
```diff
@@ -273,6 +280,74 @@ class StepFinishedEvent(BaseEvent):
...
+
+# Subtype for encrypted value
+ReasoningEncryptedValueSubtype = Literal["tool-call", "message"]
+
+
+class ReasoningStartEvent(BaseEvent):
+    """
+    Event indicating the start of a reasoning phase.
+    """
+    type: Literal[EventType.REASONING_START] = EventType.REASONING_START  # pyright: ignore[reportIncompatibleVariableOverride]
```
Comment (tylerslaton) [inline sdks/python/ag_ui/core/events.py:294]: @copilot why were we adding the pyright ignore before? I see this all over the place.

File: sdks/typescript/packages/client/src/apply/__tests__/default.reasoning.test.ts
Comment (tylerslaton) [inline sdks/typescript/packages/client/src/apply/__tests__/default.reasoning.test.ts:1]: @copilot did we miss any edge cases we should consider here? It's okay to say no.

File: sdks/typescript/packages/client/src/chunks/transform.ts [hunk 8]
```diff
@@ -9,6 +9,10 @@ import {
   ToolCallChunkEvent,
   ToolCallEndEvent,
   ToolCallStartEvent,
+  ReasoningMessageChunkEvent,
+  ReasoningMessageContentEvent,
+  ReasoningMessageEndEvent,
+  ReasoningMessageStartEvent,
```
Comment (tylerslaton) [inline sdks/typescript/packages/client/src/chunks/transform.ts:15]: Just a smell, but I notice we don't interact with the encrypted reasoning anywhere here. We do else-where. Are we missing logic that should be in here?

File: sdks/typescript/packages/client/src/middleware/__tests__/backward-compatibility-0-0-45.test.ts
Comment (tylerslaton) [inline sdks/typescript/packages/client/src/middleware/__tests__/backward-compatibility-0-0-45.test.ts:1]: @copilot did we miss any edge cases we should consider here? It's okay to say no.

File: sdks/typescript/packages/client/src/middleware/backward-compatibility-0-0-45.ts [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+  private currentMessageId: string | null = null;
+
+  override run(input: RunAgentInput, next: AbstractAgent): Observable<BaseEvent> {
+    // Reset state for each run
+    this.currentReasoningId = null;
+    this.currentMessageId = null;
+
+    return this.runNext(input, next).pipe(map((event) => this.transformEvent(event)));
+  }
+
```
Comment (tylerslaton) [inline sdks/typescript/packages/client/src/middleware/backward-compatibility-0-0-45.ts:48]: It may be nice to inform users of the transformation occurring + a way to disable the warning.

What do you think though? It could being confusing for integration partners if they don't know this is happening. Let me know.

```suggestion
private warnAboutTransformation(from: string, to: string) {
  if (process.env.SUPPRESS_TRANSFORMATION_WARNINGS) return

  console.warn(
    `AG-UI is converting ${from} to ${to}. To suppress this warning, switch over to REASONING_* events or set SUPPRESS_TRANSFORMATION_WARNINGS=true in the environment`
    )
}
```

**Comment by tylerslaton** [comment]
> @tylerslaton I've opened a new pull request, #1065, to work on those changes. Once the pull request is ready, I'll request review from you.

Well that's new, never seen it do that before 😓

File: docs/concepts/events.mdx (hunk 1 above)
Comment (AlemTuzlak) [reply docs/concepts/events.mdx:561]: added links to point to prio art

File: docs/concepts/messages.mdx (hunk 2 above)
Comment (AlemTuzlak) [reply docs/concepts/messages.mdx:120]: yes, as discussed with Markus on a call Gemini ties in reasoning to either tool-calls or normal messages or reasoning-messages. Even if you don't specify you want reasoning the llm can internally reason about the message and send back the signature without any reasoning events

File: docs/concepts/messages.mdx (hunk 3 above)
Comment (AlemTuzlak) [reply docs/concepts/messages.mdx:131]: added

File: docs/concepts/reasoning.mdx (hunk 4 above)
Comment (AlemTuzlak) [reply docs/concepts/reasoning.mdx:24]: done

File: sdks/typescript/packages/client/src/chunks/transform.ts (hunk 8 above)
Comment (AlemTuzlak) [reply sdks/typescript/packages/client/src/chunks/transform.ts:15]: I think that's correct but would love @mme to confirm this, I just pipe the event further on line 141

File: sdks/typescript/packages/client/src/middleware/backward-compatibility-0-0-45.ts (hunk 9 above)
Comment (AlemTuzlak) [reply sdks/typescript/packages/client/src/middleware/backward-compatibility-0-0-45.ts:48]: We could do that, I just followed the implementation like it was for 0.0.39, but it could be nice to tell the users they're running deprecated events, this would require docs on how to disable these logs though somewhere

File: sdks/typescript/packages/client/src/chunks/transform.ts (hunk 8 above)
Comment (mme) [reply sdks/typescript/packages/client/src/chunks/transform.ts:15]: Yes- the encrypted reasoning content would be handled on the integration level. For example, messages would be converted to LangGraph's format and handled inside of the agent framework.

File: sdks/typescript/packages/client/src/middleware/backward-compatibility-0-0-45.ts (hunk 9 above)
Comment (mme) [reply sdks/typescript/packages/client/src/middleware/backward-compatibility-0-0-45.ts:48]: I would say we should let them know how to upgrade to get rid of the warning in addition to the env variable.

File: docs/concepts/messages.mdx (hunk 2 above)
Comment (tylerslaton) [reply docs/concepts/messages.mdx:120]: Great answer, thanks guys.

**Review by ranst91 — APPROVED** [review]
